File: android/app/src/main/java/com/workapp/hafaka/data/Store.kt
package com.workapp.hafaka.data

import android.content.Context
import com.workapp.hafaka.model.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import java.io.File

/**
 * Local-first state.
 *
 * Everything lives in memory and is written to a JSON file in the app's
 * private storage. The app is fully usable with no signal, which is the
 * normal condition on location; sync is a background reconciliation on top
 * and never a precondition.
 */
class Store(
    context: Context,
    private val scope: CoroutineScope,
    filename: String = "workapp-state.json",
) {

    private val file = File(context.filesDir, filename)

    private val _state = MutableStateFlow(load())
    val state: StateFlow<AppState> = _state.asStateFlow()

    private val _syncStatus = MutableStateFlow<SyncStatus>(SyncStatus.Off)
    val syncStatus: StateFlow<SyncStatus> = _syncStatus.asStateFlow()

    private val sync = Sync(this)
    private var saveJob: Job? = null

    sealed interface SyncStatus {
        data object Off : SyncStatus
        data object Syncing : SyncStatus
        data object Offline : SyncStatus
        data class Ok(val at: Long) : SyncStatus
        data class Error(val message: String) : SyncStatus
    }

    // ------------------------------------------------------------ persistence

    private fun load(): AppState = runCatching {
        if (!file.exists()) AppState()
        else AppJson.decodeFromString<AppState>(file.readText())
    }.getOrElse { AppState() }

    /** Debounced: typing into a field shouldn't hit the disk on every keystroke. */
    private fun scheduleSave() {
        saveJob?.cancel()
        saveJob = scope.launch {
            delay(400)
            saveNow()
        }
    }

    suspend fun saveNow() = withContext(Dispatchers.IO) {
        runCatching {
            // Write-then-rename: a crash mid-write can't leave a truncated file.
            val tmp = File(file.parentFile, file.name + ".tmp")
            tmp.writeText(AppJson.encodeToString(_state.value))
            tmp.renameTo(file)
        }
    }

    private fun now() = System.currentTimeMillis().toDouble()

    private fun mutate(pushes: Boolean = true, block: (AppState) -> AppState) {
        _state.value = block(_state.value)
        scheduleSave()
        if (pushes && _state.value.settings.sync.isUsable) {
            scope.launch { sync.push() }
        }
    }

    /** Replace wholesale — sync pull and backup restore only. */
    fun replace(next: AppState) {
        _state.value = next
        scope.launch { saveNow() }
    }

    internal fun setSyncStatus(status: SyncStatus) { _syncStatus.value = status }

    // ------------------------------------------------------------ people

    val people: List<Person> get() = _state.value.people
    fun person(id: String?): Person? = id?.let { pid -> _state.value.people.firstOrNull { it.id == pid } }

    fun upsertPerson(person: Person) = mutate { s ->
        val p = person.copy(updatedAt = now())
        val i = s.people.indexOfFirst { it.id == p.id }
        s.copy(people = if (i >= 0) s.people.toMutableList().also { it[i] = p } else s.people + p)
    }

    fun deletePerson(id: String) = mutate { s ->
        s.copy(
            people = s.people.filterNot { it.id == id },
            deleted = s.deleted + (id to now()),
            // Unassign everywhere so no sheet points at a ghost.
            days = s.days.map { d ->
                d.copy(
                    slots = d.slots.filterValues { it != id },
                    calls = d.calls - id,
                    vehicles = d.vehicles.mapValues { (_, v) ->
                        if (v.driverId == id) v.copy(driverId = "") else v
                    },
                    updatedAt = now(),
                )
            },
        )
    }

    // ------------------------------------------------------------ locations

    val locations: List<Location> get() = _state.value.locations
    fun location(id: String?): Location? = id?.let { lid -> _state.value.locations.firstOrNull { it.id == lid } }

    fun upsertLocation(location: Location) = mutate { s ->
        val l = location.copy(updatedAt = now())
        val i = s.locations.indexOfFirst { it.id == l.id }
        s.copy(locations = if (i >= 0) s.locations.toMutableList().also { it[i] = l } else s.locations + l)
    }

    fun deleteLocation(id: String) = mutate { s ->
        s.copy(
            locations = s.locations.filterNot { it.id == id },
            deleted = s.deleted + (id to now()),
            days = s.days.map { if (it.locationId == id) it.copy(locationId = "") else it },
        )
    }

    // ------------------------------------------------------------ days

    val days: List<ShootDay> get() = _state.value.days.sortedBy { it.date }
    fun day(id: String?): ShootDay? = id?.let { did -> _state.value.days.firstOrNull { it.id == did } }

    /** Today if it exists, else the next upcoming, else the most recent. */
    val currentDay: ShootDay?
        get() {
            val today = HebDate.todayIso()
            val all = days
            return all.firstOrNull { it.date == today }
                ?: all.firstOrNull { it.date > today }
                ?: all.lastOrNull()
        }

    fun upsertDay(day: ShootDay) = mutate { s ->
        val d = day.copy(updatedAt = now())
        val i = s.days.indexOfFirst { it.id == d.id }
        s.copy(days = if (i >= 0) s.days.toMutableList().also { it[i] = d } else s.days + d)
    }

    fun deleteDay(id: String) = mutate { s ->
        s.copy(days = s.days.filterNot { it.id == id }, deleted = s.deleted + (id to now()))
    }

    fun editDay(id: String, block: (ShootDay) -> ShootDay) = mutate { s ->
        s.copy(days = s.days.map { if (it.id == id) block(it).copy(updatedAt = now()) else it })
    }

    fun assign(dayId: String, slot: String, personId: String?) = editDay(dayId) { d ->
        d.copy(slots = if (personId.isNullOrBlank()) d.slots - slot else d.slots + (slot to personId))
    }

    fun setCall(dayId: String, personId: String, call: CallInfo) = editDay(dayId) { d ->
        d.copy(calls = d.calls + (personId to call))
    }

    fun setCount(dayId: String, key: String, value: Int?) = editDay(dayId) { d ->
        d.copy(catering = if (value == null) d.catering - key else d.catering + (key to value))
    }

    fun roster(day: ShootDay): List<RosterEntry> = Roster.of(day, people)

    // ------------------------------------------------------------ settings

    val settings: AppSettings get() = _state.value.settings

    fun updateSettings(block: (AppSettings) -> AppSettings) =
        mutate(pushes = false) { it.copy(settings = block(it.settings)) }

    /** The whole dataset as JSON, for sharing to another device. */
    fun exportBackup(): String = AppJson.encodeToString(_state.value)

    // ------------------------------------------------------------ sync

    fun startSync() = scope.launch { sync.start() }
    fun syncNow() = scope.launch { sync.pull(); sync.push() }
    suspend fun testSync(cfg: SyncConfig): Result<Unit> = sync.test(cfg)
}

File: ios/WorkApp/Services/Store.swift
import Foundation
import Combine

/// Local-first state.
///
/// Everything is held in memory and written to a JSON file in Application
/// Support. That keeps the app fully usable with no signal, which is the
/// normal condition on location. Team sync is a background reconciliation on
/// top, never a precondition for using the app.
@MainActor
final class Store: ObservableObject {

    @Published private(set) var state = AppState()

    /// Set by Sync so the UI can show what the connection is doing.
    @Published var syncStatus: SyncStatus = .off

    enum SyncStatus: Equatable {
        case off, syncing, ok(Date), offline, error(String)
    }

    private let fileURL: URL
    private var saveTask: Task<Void, Never>?
    private lazy var sync = Sync(store: self)

    init(filename: String = "workapp-state.json") {
        let dir = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        fileURL = dir.appendingPathComponent(filename)
        load()
    }

    // MARK: Persistence

    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        if let decoded = try? JSONDecoder().decode(AppState.self, from: data) {
            state = decoded
        }
    }

    /// Debounced write — typing into a text field shouldn't hit the disk on
    /// every keystroke.
    private func scheduleSave() {
        saveTask?.cancel()
        let snapshot = state
        saveTask = Task { [fileURL] in
            try? await Task.sleep(nanoseconds: 400_000_000)
            guard !Task.isCancelled else { return }
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.withoutEscapingSlashes]
            if let data = try? encoder.encode(snapshot) {
                try? data.write(to: fileURL, options: .atomic)
            }
        }
    }

    /// Flush immediately — used when the app is backgrounded.
    func saveNow() {
        saveTask?.cancel()
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.withoutEscapingSlashes]
        if let data = try? encoder.encode(state) {
            try? data.write(to: fileURL, options: .atomic)
        }
    }

    private func mutate(pushes: Bool = true, _ block: (inout AppState) -> Void) {
        block(&state)
        scheduleSave()
        if pushes && state.settings.sync.isUsable {
            Task { await sync.push() }
        }
    }

    /// Replace wholesale — sync pull and backup restore only.
    func replace(with next: AppState) {
        state = next
        saveNow()
    }

    // MARK: People

    var people: [Person] { state.people }

    func person(_ id: String) -> Person? { state.people.first { $0.id == id } }

    func upsert(_ person: Person) {
        var p = person
        p.updatedAt = Date().timeIntervalSince1970 * 1000
        mutate { s in
            if let i = s.people.firstIndex(where: { $0.id == p.id }) { s.people[i] = p }
            else { s.people.append(p) }
        }
    }

    func deletePerson(_ id: String) {
        mutate { s in
            s.people.removeAll { $0.id == id }
            s.deleted[id] = Date().timeIntervalSince1970 * 1000
            // Unassign everywhere so no sheet points at a ghost.
            for i in s.days.indices {
                s.days[i].slots = s.days[i].slots.filter { $0.value != id }
                s.days[i].calls.removeValue(forKey: id)
                for (k, v) in s.days[i].vehicles where v.driverId == id {
                    s.days[i].vehicles[k]?.driverId = ""
                }
                s.days[i].updatedAt = Date().timeIntervalSince1970 * 1000
            }
        }
    }

    // MARK: Locations

    var locations: [Location] { state.locations }

    func location(_ id: String) -> Location? { state.locations.first { $0.id == id } }

    func upsert(_ location: Location) {
        var l = location
        l.updatedAt = Date().timeIntervalSince1970 * 1000
        mutate { s in
            if let i = s.locations.firstIndex(where: { $0.id == l.id }) { s.locations[i] = l }
            else { s.locations.append(l) }
        }
    }

    func deleteLocation(_ id: String) {
        mutate { s in
            s.locations.removeAll { $0.id == id }
            s.deleted[id] = Date().timeIntervalSince1970 * 1000
            for i in s.days.indices where s.days[i].locationId == id {
                s.days[i].locationId = ""
            }
        }
    }

    // MARK: Days

    var days: [ShootDay] { state.days.sorted { $0.date < $1.date } }

    func day(_ id: String) -> ShootDay? { state.days.first { $0.id == id } }

    /// Today if it exists, else the next upcoming, else the most recent.
    var currentDay: ShootDay? {
        let today = ShootDay.isoToday()
        let all = days
        return all.first { $0.date == today }
            ?? all.first { $0.date > today }
            ?? all.last
    }

    func upsert(_ day: ShootDay) {
        var d = day
        d.updatedAt = Date().timeIntervalSince1970 * 1000
        mutate { s in
            if let i = s.days.firstIndex(where: { $0.id == d.id }) { s.days[i] = d }
            else { s.days.append(d) }
        }
    }

    func deleteDay(_ id: String) {
        mutate { s in
            s.days.removeAll { $0.id == id }
            s.deleted[id] = Date().timeIntervalSince1970 * 1000
        }
    }

    /// Apply an edit to one day in place.
    func editDay(_ id: String, _ block: (inout ShootDay) -> Void) {
        mutate { s in
            guard let i = s.days.firstIndex(where: { $0.id == id }) else { return }
            block(&s.days[i])
            s.days[i].updatedAt = Date().timeIntervalSince1970 * 1000
        }
    }

    func assign(day dayId: String, slot: String, personId: String?) {
        editDay(dayId) { d in
            if let pid = personId, !pid.isEmpty { d.slots[slot] = pid }
            else { d.slots.removeValue(forKey: slot) }
        }
    }

    func setCall(day dayId: String, person personId: String, _ call: CallInfo) {
        editDay(dayId) { $0.calls[personId] = call }
    }

    // MARK: Derived

    /// Everyone on a day, with their call time resolved against the day's
    /// general call, sorted into arrival waves.
    func roster(for day: ShootDay) -> [RosterEntry] {
        var bySlot: [String: (slots: [String], vehicles: [String])] = [:]

        for (slot, personId) in day.slots {
            bySlot[personId, default: ([], [])].slots.append(slot)
        }
        for (vslot, v) in day.vehicles where !v.driverId.isEmpty {
            bySlot[v.driverId, default: ([], [])].vehicles.append(vslot)
        }

        return bySlot.compactMap { personId, roles -> RosterEntry? in
            guard let person = person(personId) else { return nil }
            let override = day.calls[personId]
            var call = override ?? CallInfo()
            if call.time.isEmpty { call.time = day.generalCall }
            if call.locationId.isEmpty { call.locationId = day.locationId }
            return RosterEntry(
                person: person,
                // Keep the workbook's column order, not dictionary order.
                slots: roles.slots.sorted { a, b in
                    (Sheets.all.firstIndex { $0.slot == a } ?? 0)
                        < (Sheets.all.firstIndex { $0.slot == b } ?? 0)
                },
                vehicles: roles.vehicles.sorted(),
                call: call,
                isOverride: !(override?.time.isEmpty ?? true))
        }
        .sorted {
            $0.call.time == $1.call.time
                ? $0.person.name.localizedCompare($1.person.name) == .orderedAscending
                : $0.call.time < $1.call.time
        }
    }

    // MARK: Settings

    var settings: AppSettings { state.settings }

    func updateSettings(_ block: (inout AppSettings) -> Void) {
        mutate(pushes: false) { block(&$0.settings) }
    }

    // MARK: Sync entry points

    func startSync() { Task { await sync.start() } }
    func syncNow() { Task { await sync.pull(); await sync.push() } }
    func testSync(_ cfg: SyncConfig) async -> Result<Void, SyncError> { await sync.test(cfg) }
}
